cmd/kubeadm/app/cmd/certs_test.go

```diff
@@ -243,26 +251,35 @@ func TestRunRenewCommands(t *testing.T) {
 			}
 
 			// exec renew
-			renewCmds := getRenewSubCommands(os.Stdout, tmpDir)
-			cmdtestutil.RunSubCommand(t, renewCmds, test.command, fmt.Sprintf("--cert-dir=%s", tmpDir))
-
-			// check the file is modified
-			for _, cert := range test.Certs {
-				file, err := os.Stat(filepath.Join(tmpDir, fmt.Sprintf("%s.crt", cert.BaseName)))
-				if err != nil {
-					t.Fatalf("couldn't get certificate %s: %v", cert.Name, err)
-				}
-				if ModTime[cert.Name] == file.ModTime() {
-					t.Errorf("certificate %s was not renewed as expected", cert.Name)
-				}
+			renewCmds := getRenewSubCommands(tmpDir)
+			args := fmt.Sprintf("--cert-dir=%s", tmpDir)
+			if len(test.Args) > 0 {
+				args = test.Args + " " + args
 			}
-			for _, kubeConfig := range test.KubeconfigFiles {
-				file, err := os.Stat(filepath.Join(tmpDir, kubeConfig))
-				if err != nil {
-					t.Fatalf("couldn't get kubeconfig %s: %v", kubeConfig, err)
+			err := cmdtestutil.RunSubCommand(t, renewCmds, test.command, args)
+			// certs renew doesn't support positional Args
+			if len(test.Args) > 0 {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
```

Comment:
instead of branching here, why to simply check if `err` out of `err := cmdtestutil.RunSubCommand(t, renewCmds, test.command, args)` is expected to not be non-nil?

Reply:
actually we can continue having RunSubCommand do a t.Fatal instead of returning err..

Reply:
`t.Fatalf` will exit directly iiuc, how can we assert the failure here?

Reply:
just `t.Fatalf` on `args > 1` if `cmd.Args = cobra.NoArgs` is set seems like a sufficient test.

Reply:
but it will say that testcase is failed, right? How can we let the test knows that failure is expected?

Reply:
the common way to do this it is to have `expectedError bool` in the test case struct.
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/cmd/kubeadm/app/apis/kubeadm/v1beta2/bootstraptokenstring_test.go#L72

and i guess we have to return err from cmdtestutil.RunSubCommand

Reply:
Done